File: 600/654 - Maximum Binary Tree.c
#include <limits.h>

static struct TreeNode *helper(int *nums, int from, int to) {
    int m = INT_MIN, mi = -1, i;
    struct TreeNode *ret;
    
    if (from > to)
        return NULL;
    
    for (i = from; i <= to; ++i) {
        if (nums[i] > m) {
            m = nums[i];
            mi = i;
        }
    }
    
    ret = (struct TreeNode *)malloc(sizeof(struct TreeNode));
    ret->val = m;
    ret->left = helper(nums, from, mi - 1);
    ret->right = helper(nums, mi + 1, to);
    
    return ret;
}

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     struct TreeNode *left;
 *     struct TreeNode *right;
 * };
 */
struct TreeNode* constructMaximumBinaryTree(int* nums, int numsSize) {
    return helper(nums, 0, numsSize - 1);
}

File: csharp/654.maximum-binary-tree.cs
/*
 * @lc app=leetcode id=654 lang=csharp
 *
 * [654] Maximum Binary Tree
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    
    private TreeNode ConstructMaximumBinaryTreeRecurse(int[] nums, int start, int end) {
        if (end < start) {
            return null;
        }
        int mi = start;
        for (int i = start + 1; i <= end; ++i) {
            if (nums[i] > nums[mi]) {
                mi = i;
            }
        }
        TreeNode ret = new TreeNode(nums[mi]);
        ret.left = ConstructMaximumBinaryTreeRecurse(nums, start, mi - 1);
        ret.right = ConstructMaximumBinaryTreeRecurse(nums, mi + 1, end);
        return ret;
    }
    
    public TreeNode ConstructMaximumBinaryTree(int[] nums) {
        return ConstructMaximumBinaryTreeRecurse(nums, 0, nums.Length - 1);
    }
}
// @lc code=end

